pacman/utilities/algorithm_utilities/routing_info_allocator_utilities.py

```diff
@@ -26,39 +32,542 @@ def get_edge_groups(machine_graph, traffic_type):
-                    none_continuous_groups.append(partition)
-    return (fixed_key_groups, fixed_mask_groups, fixed_field_groups,
-            flexi_field_groups, continuous_groups, none_continuous_groups)
+
+                # process the partition into the correct fields
+                _process_partition(
+                    fixed_key_groups=fixed_key_groups,
+                    shared_key_groups=shared_key_groups,
+                    fixed_mask_groups=fixed_mask_groups,
+                    fixed_field_groups=fixed_field_groups,
+                    flexi_field_groups=flexi_field_groups,
+                    continuous_groups=continuous_groups,
+                    none_continuous_groups=none_continuous_groups,
+                    partition=partition, partition_mappings=partition_mappings)
+
+    # return the set of groups
+    return (fixed_key_groups, shared_key_groups,
+            fixed_mask_groups, fixed_field_groups, flexi_field_groups,
+            continuous_groups, none_continuous_groups)
+
+
+def _process_partition(
+        fixed_key_groups, shared_key_groups, fixed_mask_groups,
+        fixed_field_groups, flexi_field_groups, continuous_groups,
+        none_continuous_groups, partition, partition_mappings):
+    """ handles the discovery of this partitions key group based off previous\
+    processed and its constraints
+
+    :param fixed_key_groups: the group of sets of fixed key partitions, where \
+    each set is a set of partitions which share the same key
+    :param shared_key_groups: the group of sets of partitions with only share \
+    key constraints, where each set is a set of partitions which share the\
+     same key
+    :param fixed_mask_groups: the group of sets of fixed masks partitions, \
+    where each set is a set of partitions which share the same key
+    :param fixed_field_groups:  the group of sets of fixed field partitions, \
+    where each set is a set of partitions which share the same key
+    :param flexi_field_groups: the group of sets of flexi field partitions, \
+    where each set is a set of partitions which share the same key.
+    :param continuous_groups: the group of sets of continuous key partitions, \
+    where each set is a set of partitions which share the same key.
+    :param none_continuous_groups: the group of sets of none continuous key \
+    partitions, where each set is a set of partitions which share the same key.
+    :param partition: The partition to place in the correct group.
+    :param partition_mappings: the mapping between partition and (set of \
+    partitions which share the same key, constraint group)
+    :rtype: None
+    """
+
+    # get types of constraints from this partition
+    (is_continuous, is_fixed_mask, is_fixed_key, is_flexi_field,
+     is_fixed_field, is_shared_key) = _check_types_of_constraints(
+        partition.constraints)
+
+    # this partition already in a share key somewhere, merge in
+    if partition in partition_mappings:
```

Comment:
I don't *think* the following code is wrong, but it is hard to follow...  A possible suggestion to make it easier is to realise that shared key is the only constraint that can group partitions.  Thus it might be easier to first group partitions and then define which of the other constraints apply to the group.

Reply:
it may be the only constraint that can add to a group, but by adding, you get the possibility of a cascade search. By handling each one separately, you (i think) reduce the amount of cycles, and allow you to consider just the addition of the constraints as they come. It also makes it faster to diagnose dodgy setups. But as i got myself scared of this code, simplification without performance reductions im happy for you to try. I'm still dripping from the wounds this gave me

Reply:
I will have a look...